receiver/prometheusreceiver/internal/metricsbuilder.go

```diff
@@ -0,0 +1,540 @@
+	},
+	Metrics: make([]*metricspb.Metric, 0),
+}
+
+type metricBuilder struct {
+	node                   *commonpb.Node
+	ts                     int64
+	hasData                bool
+	hasInternalMetric      bool
+	mc                     MetadataCache
+	metrics                []*metricspb.Metric
+	currentMetricFamily    string
+	currentMetricLabelKeys map[string]int
+	currentMetadata        *scrape.MetricMetadata
+	currentMetric          *metricspb.Metric
+	currentDpGroups        map[string][]*dataPoint
+	currentDpGOrder        map[string]int
+	currentMetadataGroups  map[string]*dataPointGroupMetadata
+	logger                 *zap.SugaredLogger
+}
+
+type dataPoint struct {
+	value    float64
+	boundary float64
+	ts       int64
+}
+
+type dataPointGroupMetadata struct {
+	ts       int64
+	ls       *labels.Labels
+	count    float64
+	hasCount bool
+	sum      float64
+	hasSum   bool
+}
+
+// newMetricBuilder creates a MetricBuilder which is allowed to feed all the datapoints from a single prometheus
+// scraped page by calling its AddDataPoint function, and turn them into an opencensus data.MetricsData object
+// by calling its Build function
+func newMetricBuilder(node *commonpb.Node, mc MetadataCache, logger *zap.SugaredLogger) *metricBuilder {
+
+	return &metricBuilder{
+		node:    node,
+		mc:      mc,
+		metrics: make([]*metricspb.Metric, 0),
+		logger:  logger,
+	}
+}
+
+// AddDataPoint is for feeding prometheus data points in its processing order
+func (b *metricBuilder) AddDataPoint(ls labels.Labels, t int64, v float64) error {
+	metricName := ls.Get(model.MetricNameLabel)
+	if metricName == "" {
+		return errMetricNameNotFound
+	} else if shouldSkip(metricName) {
+		b.hasInternalMetric = true
+		lm := ls.Map()
+		delete(lm, model.MetricNameLabel)
+		b.logger.Infow("skip internal metric", "name", metricName, "ts", t, "value", v, "labels", lm)
+		return nil
+	}
+
+	b.hasData = true
+	if err := b.initNewMetricIfNeeded(metricName, ls); err != nil {
+		return err
+	}
+
+	// update the labelKeys array, need to do it for every metrics as scrapeLoop can remove labels with empty values
+	// only when we complete the whole MetricFamily, we can get the full set of label names
+	b.updateLabelKeys(ls)
+	groupKey := dpgSignature(b.currentMetricLabelKeys, ls)
+	mg, ok := b.currentMetadataGroups[groupKey]
+	if !ok {
+		mg = &dataPointGroupMetadata{ts: t, ls: &ls}
+		b.currentMetadataGroups[groupKey] = mg
+	}
+	switch mtype := b.currentMetric.MetricDescriptor.Type; mtype {
+	case metricspb.MetricDescriptor_GAUGE_DISTRIBUTION:
+		fallthrough
+	case metricspb.MetricDescriptor_CUMULATIVE_DISTRIBUTION:
+		fallthrough
+	case metricspb.MetricDescriptor_SUMMARY:
+		if strings.HasSuffix(metricName, metricsSuffixSum) {
+			mg.sum = v
+			mg.hasSum = true
+		} else if strings.HasSuffix(metricName, metricsSuffixCount) {
+			mg.count = v
+			mg.hasCount = true
+		} else {
+			pg, ok := b.currentDpGroups[groupKey]
+			if !ok {
+				pg = make([]*dataPoint, 0, 4)
```

Comment:
It's just some number in my mind trying to preallocate a slice, there's no reason to have this value. probably better just change it to 0 to avoid confusions